fenzo-core/src/main/java/com/netflix/fenzo/ScaleDownConstraintExecutor.java

```diff
@@ -1,38 +1,92 @@
+ *         {@link ScaleDownConstraintExecutor}. This list is returned as a result.
+ *     </li>
+ * </ul>
  */
 class ScaleDownConstraintExecutor {
 
     private static final Logger logger = LoggerFactory.getLogger(ScaleDownConstraintExecutor.class);
 
-    private final List<ScaleDownConstraintEvaluator> evaluators;
+    private static final double NOT_REMOVABLE_MARKER = -1;
+
+    private final ScaleDownOrderEvaluator orderEvaluator;
+    private final List<ScaleDownConstraintEvaluator> scoringEvaluators;
 
-    ScaleDownConstraintExecutor(List<ScaleDownConstraintEvaluator> evaluators) {
-        this.evaluators = evaluators;
+    ScaleDownConstraintExecutor(ScaleDownOrderEvaluator orderEvaluator, List<ScaleDownConstraintEvaluator> scoringEvaluators) {
+        this.orderEvaluator = orderEvaluator;
+        this.scoringEvaluators = scoringEvaluators;
     }
 
-    ScaleDownOrderResolver evaluate(Collection<VirtualMachineLease> candidates) {
-        List<ScaleDownConstraintEvaluator.Result> results = new ArrayList<>();
-        evaluators.forEach(evaluator -> {
-            try {
-                results.add(evaluator.evaluate(candidates));
-            } catch (Exception e) {
-                logger.info("Ignoring scale down constraint evaluator {} due to evaluation error", evaluator.getName(), e);
+    List<VirtualMachineLease> evaluate(Collection<VirtualMachineLease> candidates) {
+        List<Set<VirtualMachineLease>> fixedOrder = orderEvaluator.evaluate(candidates);
+
+        List<VirtualMachineLease> scaleDownOrder = scoringEvaluators.isEmpty()
+                ? fixedOrder.stream().flatMap(Set::stream).collect(Collectors.toList())
+                : fixedOrder.stream().flatMap(this::groupEvaluator).collect(Collectors.toList());
+
+        if (logger.isDebugEnabled()) {
+            List<String> hosts = scaleDownOrder.stream().map(VirtualMachineLease::hostname).collect(Collectors.toList());
+            logger.debug("Evaluated scale down order: {}", hosts);
+        }
+
+        return scaleDownOrder;
+    }
+
+    private Stream<VirtualMachineLease> groupEvaluator(Set<VirtualMachineLease> groupCandidates) {
+        Map<VirtualMachineLease, Double> scores = new HashMap<>();
+
+        scoringEvaluators.forEach(e -> {
+            Optional<? super Object> optionalContext = Optional.empty();
+            for (VirtualMachineLease lease : groupCandidates) {
+                double currentScore = scores.getOrDefault(lease, 0.0);
+                if (currentScore != NOT_REMOVABLE_MARKER) {
+                    ScaleDownConstraintEvaluator.Result result = e.evaluate(lease, optionalContext);
+                    double newScore = result.getScore();
+                    if (newScore == 0) {
+                        scores.put(lease, NOT_REMOVABLE_MARKER);
+                    } else {
+                        scores.put(lease, currentScore + newScore);
```

Comment:
Yes, there are no weight there right now. I assumed they are all equal, but I agree adding weights makes sense, and is not complicated. I will change that.